docs/apache-airflow/authoring-and-scheduling/datasets.rst

```diff
@@ -224,6 +224,29 @@ If one dataset is updated multiple times before all consumed datasets have been
 
     }
 
+Attaching extra information to an emitting Dataset Event
+--------------------------------------------------------
+
+.. versionadded:: 2.10.0
+
+A task with a dataset outlet can optionally attach extra information before it emits a dataset event. This is different
+from `Extra information on Dataset`_. Extra information on a dataset statically describes the entity pointed to by the dataset URI; extra information on the *dataset event* instead should be used to annotate the triggering data change, such as how many rows in the database are changed by the update, or the date range covered by it.
+
+The easiest way to attach extra information to the dataset event is by accessing ``dataset_events`` in a task's execution context:
+
+.. code-block:: python
+
+    example_s3_dataset = Dataset("s3://dataset/example.csv")
+
+
+    @task(outlets=[example_s3_dataset])
+    def write_to_s3(*, dataset_events):
+        df = ...  # Get a Pandas DataFrame to write.
+        # Write df to dataset...
+        dataset_events[example_s3_dataset].extras = {"row_count": len(df)}
+
+This can also be done in classic operators by either subclassing the operator and overriding ``execute``, or by supplying a pre- or post-execution function.
```

Comment:
Are you sure that attaching extra is also possible in `post_execution` hook? Sure this is not executed after the event has been emitted?

Reply:
I believe so, the callback is executed in `_execute_task_with_callbacks`, while dataset events are emitted in _register_dataset_changes`. Both functions are called in `_run_raw_task` and the callback is run earlier. I should add a test for this.

Reply:
Yup, just added a test case, this works.

Reply:
> This can also be done in classic operators by either subclassing the operator and overriding ``execute``, or by supplying a pre- or post-execution function.

Would it be possible to achieve the same behaviour _without_ the constraint of subclassing classical operators?

Feature adoption becomes seamless when things work out-of-the-box! The present implementation requires users to subclass classical operators to take benefit of the feature. 

I am unsure if the above concern is valid, but figured I'll bring it to your attend and gauge your thoughts. (Please correct me if that's not the case)

Reply:
See attached test for an example. No subclassing needed.

Reply:
Sorry, I hope I don't sound dumb - Will this work for operators like `AWSGlueOperator` or `DatabricksSubmitRunOperator` - which do not accept user-provided lambdas?

```py
def _write2_post_execute(context, _):
    context["dataset_events"]["test_outlet_dataset_extra_2"].extra = {"x": 1}

BashOperator(
    task_id="write2",
    bash_command=":",
    outlets=Dataset("test_outlet_dataset_extra_2"),
    post_execute=_write2_post_execute,
)
```

As I understand from the above code, we are relying on `context` variable inside the `_write2_post_execute` lambda to publish the _extras_. Not all operators accept user-provided lambdas. (case in point: `AWSGlueOperator`, `DatabricksSubmitRunOperator`)

I intend to adopt this functionality once released, hence the question.

Reply:
The `post_execute` hook is a core Airflow feature and should be accepted by all operators. Please report if an operator doesn’t handle this correctly; it should be considered a bug.